src/sentry/sentry_metrics/consumers/indexer/batch.py

```diff
@@ -241,19 +254,24 @@ def reconstruct_messages(
                             exceeded_org_quotas += 1
                         continue
 
-                    if new_v is None:
-                        metadata = bulk_record_meta[org_id].get(v)
-                        if (
-                            metadata
-                            and metadata.fetch_type_ext
-                            and metadata.fetch_type_ext.is_global
-                        ):
-                            exceeded_global_quotas += 1
+                    value_to_write = v
+                    if self._should_index_tag_values():
+                        new_v = mapping[org_id][v]
+                        if new_v is None:
+                            metadata = bulk_record_meta[org_id].get(v)
+                            if (
+                                metadata
+                                and metadata.fetch_type_ext
+                                and metadata.fetch_type_ext.is_global
+                            ):
+                                exceeded_global_quotas += 1
+                            else:
+                                exceeded_org_quotas += 1
+                            continue
                         else:
-                            exceeded_org_quotas += 1
-                        continue
+                            value_to_write = new_v
 
-                    new_tags[str(new_k)] = new_v
+                    new_tags[str(new_k)] = value_to_write
```

Comment:
So this is changing the format. 
Do you plan to make changes to the Snuba consumer to detect what it has to fill depending on the type ? 
I suggest we do something more robust and have two separate fields, one for the indexed value and one for the raw value.
This means that the snuba consumer can just pick the right column

Reply:
My intention was to use the single `tags` field right now with values which can be both `ints` and `strings`. Supporting multiple types is only needed for migrating to use `strings`. Once we migrate to using `strings` we no longer need to support `ints`. That way the protocol between release health and gen-metrics doesn't differ much except for types. So the protocol would look something like this

Now - tags: {ints: ints}
During Migration - tags: {ints: ints | strs}
After Migration - tags: {ints: strs}

In the end we just keep using the tags field.

Reply:
I am ok in reusing the same field, though, as I expressed in the Snuba PR, I would suggest you also add a `mode` field to tell the consumer what to expect. That is a bit more robust than inferring the mode (indexed vs not) from the type of the first element in the tags column.

Also, the migraiton is only concerning performance. I don't think we have a reason to migrate release health. So the support for indexed tags should stay.

Reply:
This is a case of protocol change which is generally captured by fields like `version`

So I added a `version` field which is added to the payload if tag values are strings. The value is set to 2 in that case. I was originally trying to add a `version` = 1 if strings are indexed and `version` = 2 if strings are not indexed. But that would cause the release-health payload to also change and require a bunch of tests to be fixed that are asserting for exact payload.

The consumer would have a check to see if the `version` field exists and then take the appropriate action.

Going forward we should be able to modify the version field and make payload changes in a less intrusive manner

Reply:
> This is a case of protocol change which is generally captured by fields like version

Minor semantic nit. Would not block for this. This is not really a version. Version imply that 2 is more recent/better than 1 and kind of implies a migration path. 
Indexing values vs not are just two protocols that will coexist as they are useful in different scenarios.